dell/src/main/java/org/apache/iceberg/dell/ecs/EcsTableOperations.java

```diff
@@ -0,0 +1,111 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.dell.ecs;
+
+import java.util.Map;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+public class EcsTableOperations extends BaseMetastoreTableOperations {
+
+  public static final String ICEBERG_METADATA_LOCATION = "iceberg_metadata_location";
+
+  private final String tableName;
+  private final FileIO fileIO;
+  private final EcsCatalog catalog;
+  private final EcsURI tableObject;
+
+  /**
+   * Cached E-Tag for CAS commit
+   *
+   * @see #doRefresh() when reset this field
+   * @see #doCommit(TableMetadata, TableMetadata) when use this field
+   */
+  private String eTag;
+
+  public EcsTableOperations(String tableName, EcsURI tableObject, FileIO fileIO, EcsCatalog catalog) {
+    this.tableName = tableName;
+    this.tableObject = tableObject;
+    this.fileIO = fileIO;
+    this.catalog = catalog;
+  }
+
+  @Override
+  protected String tableName() {
+    return tableName;
+  }
+
+  @Override
+  public FileIO io() {
+    return fileIO;
+  }
+
+  @Override
+  protected void doRefresh() {
+    String metadataLocation;
+    if (!catalog.objectMetadata(tableObject).isPresent()) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException("Metadata object %s is absent", tableObject);
+      } else {
+        metadataLocation = null;
+      }
+    } else {
+      EcsCatalog.Properties metadata = catalog.loadProperties(tableObject);
+      this.eTag = metadata.eTag();
+      metadataLocation = metadata.content().get(ICEBERG_METADATA_LOCATION);
+      Preconditions.checkNotNull(metadataLocation,
+          "Can't find location from table metadata %s", tableObject);
+    }
+
+    refreshFromMetadataLocation(metadataLocation);
+  }
+
+  @Override
+  protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+    if (base == null) {
+      // create a new table, the metadataKey should be absent
+      if (!catalog.putNewProperties(tableObject, buildProperties(newMetadataLocation))) {
```

Comment:
Is the Dell EMC's `putObject`  atomic ?  In this [putNewProperties](https://github.com/apache/iceberg/pull/4221/files#diff-d7308f132665917b84cff30f2688171a11a64e951fd9ed995e3040d0463cda51R465-R471),  I think the `putObject` implementation will check the existence of the original key, if the key is present, then it will throw a `PreconditionFailed`, otherwise write the content into the backend storage.

The key thing is :  the two operations ( check existence of key and write the content ) should be atomic in a single transaction. Otherwise, assume there are two concurrent writers A and B,  both A and B pass the existence check, and then try to overwrite the content to the same key.  Finally, the result will be unknown because we don't know which one has overwrote the other one.

If the atomicity is not be guaranteed,  then we can not use this `putObject` to commit the transaction. Instead, we are recommended to use an external table lock service.

Reply:
Yes. It's atomic when specified `If-None-Match`. It's an additional extension of S3 API.